[hop][be] add utils for more comprehensive input alias and mutation

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #145298

This PR implements  the idea of checking input mutations through tensor version and check aliasing via storage  from @zou3519. Previously, we rely on whether there's a in place op that takes placeholder input, which doesn't take views into account.

When writing the PR, I also noticed a bug in previous input mutation checking logic: we were checking the whether there are operators functionalized_f where all the mutating ops have been replaced so we won't be able to detect any thing.

ydwu4 committed:
[hop][be] add utils for more comprehensive input alias and mutation

[ghstack-poisoned]

ydwu4 committed:
Update on "[hop][be] add utils for more comprehensive input alias and mutation"




[ghstack-poisoned]

ydwu4 committed:
Update on "[hop][be] add utils for more comprehensive input alias and mutation"


This PR implements  the idea of checking input mutations through tensor version and check aliasing via storage  from zou3519. Previously, we rely on whether there's a in place op that takes placeholder input, which doesn't take views into account.

When writing the PR, I also noticed a bug in previous input mutation checking logic: we were checking the whether there are operators functionalized_f where all the mutating ops have been replaced so we won't be able to detect any thing.

[ghstack-poisoned]

zou3519 commented on torch/_higher_order_ops/utils.py:
```diff
@@ -546,3 +538,62 @@ def validate_subgraph_args_types(lifted_args: Union[tuple[Any, ...], list[Any]])
+            if fake_mode is not None
+            else nullcontext()
+        )
+        with torch.inference_mode(False), fake_ctx, ignore_fresh_unbacked:
+            outputs = gm(*cloned)
```
Do you have a test for this?

Also, I think the cloned tensors need to be closed when inference_mode is off. Otherwise, they may not be eligible to have version counters (tensors created under inference mode do not have version counters)

ydwu4 replied:
yeah, added tests.

zou3519 commented on torch/_higher_order_ops/utils.py:
```diff
@@ -546,3 +538,62 @@ def validate_subgraph_args_types(lifted_args: Union[tuple[Any, ...], list[Any]])
+
+        # Clone the fake args to avoid mutating the original fake args
+        cloned = [
+            arg.clone() if isinstance(arg, torch.Tensor) else arg for arg in fake_args
```
use a clone_preserving_strides

zou3519 commented on torch/_higher_order_ops/utils.py:
```diff
@@ -546,3 +538,62 @@ def validate_subgraph_args_types(lifted_args: Union[tuple[Any, ...], list[Any]])
+            for i, out in enumerate(outputs)
+            if isinstance(out, torch.Tensor)
+        }
+        out_alias_map = {
+            i: out_storage_map[_tensor_storage(out)]
+            for i, out in enumerate(outputs)
+            if isinstance(out, torch.Tensor)
+            and _tensor_storage(out) in out_storage_map
+            and out_storage_map[_tensor_storage(out)] != i
+        }
```
this is a map of outputs aliasing outputs right?

ydwu4 replied:
change the var name to be more explicit.

zou3519 commented on torch/_higher_order_ops/utils.py:
```diff
@@ -546,3 +538,62 @@ def validate_subgraph_args_types(lifted_args: Union[tuple[Any, ...], list[Any]])
+            and _tensor_storage(out) in out_storage_map
+            and out_storage_map[_tensor_storage(out)] != i
+        }
+        inp_alias_map = {
+            i: out_storage_map[_tensor_storage(inp)]
+            for i, inp in enumerate(cloned)
+            if isinstance(inp, torch.Tensor) and _tensor_storage(inp) in out_storage_map
+        }
```
This is a map of inputs aliasing outputs right?

We should also ban inputs aliasing any inputs

ydwu4 replied:
yeah, also added inp_inp_alias and a test.

zou3519 commented on torch/_higher_order_ops/utils.py:
```diff
@@ -546,3 +538,62 @@ def validate_subgraph_args_types(lifted_args: Union[tuple[Any, ...], list[Any]])
+        """
+
+        def _tensor_version(t) -> Optional[int]:
+            if isinstance(t, FakeTensor):
+                return t._version
+            return None
```
Can you assert that all Tensors are FakeTensors? Otherwise this function is wrong

zou3519 reviewed (commented):
seems reasonable, we should add a test for the inference_mode case

ydwu4 committed:
Update on "[hop][be] add utils for more comprehensive input alias and mutation"


This PR implements  the idea of checking input mutations through tensor version and check aliasing via storage  from zou3519. Previously, we rely on whether there's a in place op that takes placeholder input, which doesn't take views into account.

When writing the PR, I also noticed a bug in previous input mutation checking logic: we were checking the whether there are operators functionalized_f where all the mutating ops have been replaced so we won't be able to detect any thing.

[ghstack-poisoned]

ydwu4 committed:
Update on "[hop][be] add utils for more comprehensive input alias and mutation"


This PR implements  the idea of checking input mutations through tensor version and check aliasing via storage  from zou3519. Previously, we rely on whether there's a in place op that takes placeholder input, which doesn't take views into account.

When writing the PR, I also noticed a bug in previous input mutation checking logic: we were checking the whether there are operators functionalized_f where all the mutating ops have been replaced so we won't be able to detect any thing.

[ghstack-poisoned]

zou3519 reviewed (approved):
thank you

ydwu4 commented:
@pytorchbot merge -i
